bookserver/crud.py

```diff
@@ -494,3 +580,189 @@ async def create_user_chapter_progress_entry(
+        rslogger.debug(f"{res=}")
+        return QuestionValidator.from_orm(res.scalars().first())
+
+
+auto_gradable_q = [
+    "clickablearea",
+    "mchoice",
+    "parsonsprob",
+    "dragndrop",
+    "fillintheblank",
+]
+
+
+async def fetch_matching_questions(request_data: schemas.SelectQRequest) -> List[str]:
+
+    if request_data.questions:
+        questionlist = request_data.questions.split(",")
+        questionlist = [q.strip() for q in questionlist]
+    elif request_data.proficiency:
+        prof = request_data.proficiency
+
+        where_clause = (Competency.competency == prof) & (
+            Competency.question == Question.id
+        )
+        if request_data.primary:
+            where_clause = where_clause & (Competency.is_primary == True)  # noqa E712
+        if request_data.min_difficulty:
+            where_clause = where_clause & (
+                Question.difficulty >= float(request_data.min_difficulty)
+            )
+        if request_data.max_difficulty:
+            where_clause = where_clause & (
+                Question.difficulty <= float(request_data.max_difficulty)
+            )
+        if request_data.autogradable:
+            where_clause = where_clause & (
+                (Question.autograde == "unittest")
+                | Question.question_type.in_(auto_gradable_q)
+            )
+        query = select(Question).where(where_clause)
+
+        async with async_session() as session:
+            res = await session.execute(query)
+            rslogger.debug(f"{res=}")
+            questionlist = [q.name for q in res]
+
+    return questionlist
+
+
+async def fetch_assignment_question(assignment_name: str, question_name: str):
+
+    query = select(AssignmentQuestion).where(
+        (Assignment.name == assignment_name)
+        & (Assignment.id == AssignmentQuestion.assignment_id)
+        & (AssignmentQuestion.question_id == Question.id)
+        & (Question.name == question_name)
+    )
+
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return AssignmentQuestionValidator.from_orm(res.scalars().first())
+
+
+async def fetch_user_experiment(sid: str, ab_name: str) -> int:
+    query = (
+        select(UserExperiment.exp_group)
+        .where((UserExperiment.sid == sid) & (UserExperiment.experiment_id == ab_name))
+        .order_by(UserExperiment.id)
+    )
+    async with async_session() as session:
+        res = await session.execute(query)
+        r = res.scalars().first()
+        rslogger.debug(f"{r=}")
+        return r
+
+
+async def create_user_experiment_entry(sid: str, ab: str, group: int):
+    new_ue = UserExperiment(sid=sid, exp_group=group, experiment_id=ab)
+    async with async_session.begin() as session:
+        session.add(new_ue)
+    return UserExperimentValidator.from_orm(new_ue)
+
+
+async def fetch_viewed_questions(sid, questionlist) -> List[str]:
```

Comment:
Nitpick: missing types

Reply:
Docs: the return value is a list of div_ids for questions that were viewed? Why does the existence of a sid entry imply they were viewed?